usecases/cluster/delegate.go

```diff
@@ -0,0 +1,247 @@
+		return nil, err
+	}
+	_, err = buf.Write([]byte(d.Node))
+	return buf.Bytes(), err
+}
+
+func (d *spaceMsg) unmarshal(data []byte) error {
+	rd := bytes.NewReader(data)
+	if err := binary.Read(rd, binary.BigEndian, &d.header); err != nil {
+		return err
+	}
+	if err := binary.Read(rd, binary.BigEndian, &d.DiskUsage); err != nil {
+		return err
+	}
+	d.Node = string(data[len(data)-rd.Len():])
+	return nil
+}
+
+// delegate implements the memberList delegate interface
+type delegate struct {
+	Name     string
+	dataPath string
+	sync.Mutex
+	diskUsage func(path string) (DiskUsage, error)
+	Cache     map[string]NodeInfo
+}
+
+// init must be called first to initialize the cache
+func (d *delegate) init() error {
+	d.Cache = make(map[string]NodeInfo, 32)
+	d.diskUsage = diskSpace
+	space, err := diskSpace(d.dataPath)
+	if err != nil {
+		return fmt.Errorf("disk_space: %w", err)
+	}
+
+	d.set(d.Name, NodeInfo{space, time.Now().UnixMilli()}) // cache
+	return nil
+}
+
+// NodeMeta is used to retrieve meta-data about the current node
+// when broadcasting an alive message. It's length is limited to
+// the given byte size. This metadata is available in the Node structure.
+func (d *delegate) NodeMeta(limit int) (meta []byte) {
+	return nil
+}
+
+// LocalState is used for a TCP Push/Pull. This is sent to
+// the remote side in addition to the membership information. Any
+// data can be sent here. See MergeRemoteState as well. The `join`
+// boolean indicates this is for a join instead of a push/pull.
+func (d *delegate) LocalState(join bool) []byte {
+	prv, _ := d.get(d.Name) // value from cache
+	var (
+		info NodeInfo = prv
+		err  error
+	)
+	// renew cached value if ttl expires
+	if prv.Available == 0 || time.Since(time.UnixMilli(prv.LastTimeMilli)) >= _ProtoTTL {
+		info.DiskUsage, err = d.diskUsage(d.dataPath)
+		if err != nil {
+			return nil
+		}
+		info.LastTimeMilli = time.Now().UnixMilli()
+	}
+
+	if prv.LastTimeMilli != info.LastTimeMilli {
+		d.set(d.Name, info) // cache new value
+	}
+
+	x := spaceMsg{
+		header{OpCode: _OpCodeDisk, ProtoVersion: _ProtoVersion},
+		info.DiskUsage,
+		d.Name,
+	}
+	bytes, err := x.marshal()
+	if err != nil {
+		return nil
+	}
+	return bytes
+}
+
+// MergeRemoteState is invoked after a TCP Push/Pull. This is the
+// state received from the remote side and is the result of the
+// remote side's LocalState call. The 'join'
+// boolean indicates this is for a join instead of a push/pull.
+func (d *delegate) MergeRemoteState(data []byte, join bool) {
+	var x spaceMsg
+	if err := x.unmarshal(data); err != nil || x.Node == "" {
+		return
+	}
+	info := NodeInfo{x.DiskUsage, time.Now().UnixMilli()}
+	d.set(x.Node, info)
+}
+
+func (d *delegate) NotifyMsg(data []byte) {}
+
+func (d *delegate) GetBroadcasts(overhead, limit int) [][]byte {
+	return nil
+}
+
+// get returns info about about a specific node in the cluster
+func (d *delegate) get(node string) (NodeInfo, bool) {
+	d.Lock()
+	defer d.Unlock()
```

Comment:
Can we use `RWMutex` for `delegate` so we can `RLock` here instead?

Reply:
In terms of performance Mutex is better. RWMutex makes more sense when you have large number of reader